moved rc.mavlink to the boards optional rc additions (now it's called…

garfieldG opened the pull request:
… rc.board_mavlink)

**Describe problem solved by the proposed pull request**
I suggest to stop using the general mavlink startup script and start using board specific mavlink startup scripts.
today, rc.mavlink only tries to open mavlink channel over usb, the problem is not all boards are in need of mavlink over usb.
another thing that it can handle is the opening of mavlink over ethernet.
i think it's a more elegant solution than checking for specific boards on the general script and for costume and future boards.

dagar commented:
It makes sense why you did this, but I'm wondering if it pushes too many of the specifics of a USB mavlink connection into each board.

What do you think about pushing these USB defaults into mavlink itself? If USB then default to the high rate (-r 800000), config mode (-m config), with ftp enabled (-x).

Then having a simple `mavlink start -d /dev/ttyACM0` per board is a bit more palatable. 

It then probably also makes sense to move sercon per board. https://github.com/PX4/Firmware/blob/a26865c2792e37105b79e5767cd7aac395878953/ROMFS/px4fmu_common/init.d/rcS#L76

dagar commented on boards/av/x-v1/init/rc.board_mavlink:
```diff
@@ -1,8 +1,10 @@
+#------------------------------------------------------------------------------
+
+# Start MAVLink on the USB port
+mavlink start -r 800000 -d /dev/ttyACM0 -m config -x
```
This board doesn't have USB.

garfieldG commented:
@dagar 
Thanks for your review!
i agree with you, i'll do it later on today and push it.

garfieldG commented:
@dagar 
regarding the sercon - it makes more sense to move it per board, i'm just not sure it should be in mavlink script, what do you think? i think that board_extras might be the right fit, but we need to move it to be called previous to board_mavlink.

garfieldG commented:
@dagar 
just pushed the new commit.

regarding the sercon - didn't push it yet. 
i do think i should be board specific, but mavlink is not the only use.
we might want to use the usb as serial device for console and not mavlink.
i do think it shouldn't be called from rc.board_mavlink, but i can think about two possibilities :
-maybe from rc.board_extras and then to move board_extras call before board_mavlink call (in that case the sercon call will not be so at the start and board_extras will be moved before mavlink)
-new variable defaults to no in rcS and changed to yes in rc.board_default only in boards that support serial usb. the benefit is that we don't change the order of rcS as much as in the previous suggestion, but on the other hand a little bit more flash.

dagar commented:
Ok, let's leave sercon alone for now. What's been tested here? I recall there was a race condition related to mavlink usb that I'm not sure was ever resolved. Let me see if I can track that down to see if there's anything else we need to worry about here.

Jenkins is failing due to a minor whitespace issue. http://ci.px4.io:8080/blue/organizations/jenkins/PX4%2FFirmware/detail/PR-12952/2/pipeline
![image](https://user-images.githubusercontent.com/84712/64924787-44373800-d7b6-11e9-853a-4d8aa81b8fc4.png)

garfieldG commented:
@dagar 
fixed the style issue and squashed all to one commit.


> I recall there was a race condition related to mavlink usb that I'm not sure was ever resolved

are you referring to https://github.com/PX4/Firmware/issues/12201?
it has been fixed [here](https://github.com/PX4/Firmware/pull/12428)

dagar commented:
> are you referring to #12201?
> it has been fixed [here](https://github.com/PX4/Firmware/pull/12428)

Thanks, that's the one. Good to know it was merged.

dagar commented on src/modules/mavlink/mavlink_main.cpp:
```diff
@@ -1898,6 +1895,15 @@ Mavlink::task_main(int argc, char *argv[])
 			set_protocol(SERIAL);
+
+			/* USB serial is indicated by /dev/ttyACM0*/
+			if (strcmp(_device_name, "/dev/ttyACM0") == OK || strcmp(_device_name, "/dev/ttyACM1") == OK) {
```
Won't this be problematic for anyone actually trying to set different values for the mavlink usb instance? They'll be dependent on specifying the device before other arguments.

garfieldG replied:
@dagar 
yeah it will, you will have to specify the different values after the device.
i intentionally didn't changed the defaults [value](https://github.com/PX4/Firmware/blob/dea5afd327ccfa747bf198da2360182b9131d809/src/modules/mavlink/mavlink_main.cpp#L1840) because there's a reason why they were set for the first place, e.g. [here](https://github.com/PX4/Firmware/blob/dea5afd327ccfa747bf198da2360182b9131d809/src/modules/mavlink/mavlink_main.cpp#L2074) and they serve non usb devices as defaults.
we want different defaults just for usb. 
as i'm looking at it we a 3 options :
1. leave as is.
2. after getopt function to do checks to determine if new values have been specified or use default ones, e.g. if datarate is 0 and device is usb.
3. understand that maybe we try to exclude the mavlink over usb case, something that maybe we shouldn't do. as i think about it further more you always have to specify the different values for mavlink on the start of new mavlink stream, why is the usb case should be any different? if we do it for all other mavlink instances why not just for the usb?

another thing, on options 1 and 2 ftp will always be enabled on usb, if we don't want it to always be enabled option 3 is better because there's no default for ftp to be on and it's the users choice.

dagar replied:
I'd say give 2 a try, and if it makes a bigger mess we go with it as is (option 1.) I'm mainly trying to anticipate the case where someone inevitably wants USB with a different mode and wastes a day debugging.

garfieldG replied:
@dagar 
pushed the latest changes (squashed so there would be only one commit).
as you can see, i changed the usb check to use variable instead of mode. i've noticed that there seem to be a bit of confusion of what to use to determine if a connection is usb or not, device type (ACM) or mode (config), all that while there is a variable exactly for that (_is_usb_uart). 
further more, https://github.com/PX4/Firmware/pull/12428 fix won't work if mode is not config.

dagar commented:
@PX4/testflights can you give this a quick try on the bench with one or two vehicles? Make sure auto-connecting to QGC via USB still seems the same. Powering up with and without USB connected, disconnecting USB once already powered, etc.

dannyfpv commented:
tested on pixhawk4 v5 f-450
bench test
no issues connecting via usb cable moving all the rc inputs
no issues connecting via telemetry moving all the rc inputs
log:
https://review.px4.io/plot_app?log=70991c1b-dac1-4768-8273-05995a00758d

jorge789 commented:
### Tested on PixRacer V4:

Note:
It was ensured that the vehicle connected with QGC with and without the USB cable.

Log:
https://review.px4.io/plot_app?log=ee1dc100-686e-43c5-9194-365f6e0d8331

https://review.px4.io/plot_app?log=7c7658a3-8bc4-4996-823b-a2fa1ca949c4

Junkim3DR commented:
### Tested on Pixhawk 4mini v5

**Test**
Bench Test

**Notes**
No issues noted, vehicle connected to QGC through USB cable and Telemetry as expected.

**Log**
https://review.px4.io/plot_app?log=8df9f1dc-4296-498c-9a31-a844745d14cc


### Tested on NXP FMUK66 v3

**Test**
Bench Test

**Notes**
No issues noted, vehicle connected to QGC through USB cable and Telemetry as expected.

**Log**
https://review.px4.io/plot_app?log=e5357839-f182-4727-8f03-b8a2fc1813a1

dagar commented:
Thanks everyone!
